### PR TITLE
WIP: PiT support for versioned stores

### PR DESCRIPTION
**Pull request by julio-lopez**
- [x] store-independent interfaces and functionality
- [x] S3 implementation
- [ ] basic unit tests

Flag wiring in a separate PR (#230)

### CODE REVIEW

File: repo/blob/pit/pit.go [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+		return nil
+	})
+	if err != nil {
+		return errors.Wrapf(err, "could not list blob versions at time %s", s.pointInTime)
+	}
+
+	// process last blob
+	if v, found := newestAtUnlessDeleted(vs, s.pointInTime); found {
+		if err := cb(v.Metadata); err != nil {
+			return err
```
Comment (redgoat650) [inline repo/blob/pit/pit.go:51]: Just an observation that if the callback throws an error here, it'll behave differently, i.e. have a reduced wrapped-context, compared to if it's thrown in the `ListBlobVersions` call above. Seems ideally the error behavior should be identical whether the callback hits an error on the last blob or not.

From reading the code, looks like a callback error would wrap `"callback failed for <key>"` and `"could not list blob versions at time <point-in-time>"`, unless it was called here on the last blob. An error in the last blob's CB won't include that extra info.

File: repo/blob/pit/pit.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+	if v, found := newestAtUnlessDeleted(vml, s.pointInTime); found {
+		return v, nil
+	}
+
+	return VersionMetadata{}, blob.ErrBlobNotFound
+}
+
+func newestAtUnlessDeleted(vs []VersionMetadata, t time.Time) (v VersionMetadata, found bool) {
+	vs = getOlderThan(vs, t)
```
Comment (redgoat650) [inline repo/blob/pit/pit.go:98]: This `getOlderThan` call seems unneeded when `newestAtUnlessDeleted` is called in `getMetadata`, yeah? At this point all versions in `v` are already earlier than `s.pointInTime` after the same filter in the `GetBlobVersions` callback.

At first it seemed like redundant work, but on rereading, it should always return in the first loop of `getOlderThan`, so doesn't add much overhead

**Review by redgoat650 — APPROVED** [review]
Looks nice, @julio-lopez . Approving, but left a few observations for discussion.

File: repo/blob/s3/s3_options.go [hunk 3]
```diff
@@ -1,5 +1,9 @@
 package s3
 
+import (
+	"time"
+)
```
Comment (redgoat650) [inline repo/blob/s3/s3_options.go:5]: To match with the other PR's change? 
```suggestion
import "time"
```
https://github.com/kastenhq/kopia/pull/230

**Comment by julio-lopez** [comment]
@redgoat650 PTAL at 6caf2e7

**Comment by julio-lopez** [comment]
superseded by #240

**Pull request closed without merging**